server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,459 @@
+                    + "] not eligible to execute decommission request. Will retry until timeout."
+            );
+        }
+    }
+
+    private void excludeDecommissionedClusterManagerNodesFromVotingConfig(Set<DiscoveryNode> clusterManagerNodesToBeDecommissioned) {
+        Set<String> clusterManagerNodesNameToBeDecommissioned = clusterManagerNodesToBeDecommissioned.stream()
+            .map(DiscoveryNode::getName)
+            .collect(Collectors.toSet());
+
+        Set<VotingConfigExclusion> currentVotingConfigExclusions = clusterService.getClusterApplierService()
+            .state()
+            .coordinationMetadata()
+            .getVotingConfigExclusions();
+        Set<String> excludedNodesName = currentVotingConfigExclusions.stream()
+            .map(VotingConfigExclusion::getNodeName)
+            .collect(Collectors.toSet());
+
+        // check if the to-be-excluded nodes are excluded. If yes, we don't need to exclude them again
+        if (clusterManagerNodesNameToBeDecommissioned.size() == 0
+            || (clusterManagerNodesNameToBeDecommissioned.size() == excludedNodesName.size()
+                && excludedNodesName.containsAll(clusterManagerNodesNameToBeDecommissioned))) {
+            return;
+        }
+        // send a transport request to exclude to-be-decommissioned cluster manager eligible nodes from voting config
+        decommissionController.excludeDecommissionedNodesFromVotingConfig(
+            clusterManagerNodesNameToBeDecommissioned,
+            new ActionListener<Void>() {
+                @Override
+                public void onResponse(Void unused) {
+                    logger.info(
+                        "successfully removed decommissioned cluster manager eligible nodes [{}] from voting config ",
+                        clusterManagerNodesToBeDecommissioned.toString()
+                    );
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.debug(
+                        new ParameterizedMessage("failure in removing decommissioned cluster manager eligible nodes from voting config"),
+                        e
+                    );
+                }
+            }
+        );
+    }
+
+    /**
+     * Registers new decommissioned attribute metadata in the cluster state with {@link DecommissionStatus#DECOMMISSION_INIT}
+     * <p>
+     * This method can be only called on the cluster-manager node. It tries to create a new decommissioned attribute on the cluster manager
+     * and if it was successful it adds new decommissioned attribute to cluster metadata.
+     * <p>
+     * This method would only be executed on eligible cluster manager node
+     *
+     * @param decommissionAttribute register decommission attribute in the metadata request
+     * @param listener              register decommission listener
+     */
+    private void registerDecommissionAttribute(
```

Comment:
Can we simplify this to use `DecommissionController#updateMetadataWithDecommissionStatus` instead passing specific listeners

Reply:
There are some fundamental difference while registering it with INIT or updating the status in general. Although it can be reused but I prefer to keep both separate for simplicity and clean code